common/lib/xmodule/xmodule/capa_base.py

```diff
@@ -1216,6 +1216,7 @@ def submit_problem(self, data, override_time=False):
                 }
 
         try:
+            self.lcp.context['attempt'] = self.attempts + 1
```

Comment:
This is just weird to me. Why do we assign the `attempt` value to the `lcp.context` before we updated `self.attempts` counter on line 1221?
Does `self.lcp.grade_answers` use this key? if it does, why not just move the `self.attempts = self.attempts + 1` line up?

Reply:
The whole point of this is to attach the current attempt number to `self.lcp.context` so that a custom python grader called from `self.lcp.grade_answers` can access it. Note that if the grader raises an exception (such as unable to parse input etc), the attempt number `self.attempts` is not incremented. So, on the first attempt, the attempt counter starts at 0, and only increments if a non-exception-raising attempt occurs. However, for the purpose of the grader, that first attempt should be treated as attempt one. Similarly for later attempts.

Reply:
@jolyonb I see. Then can you add some comment here to spell out that `self.lcp.context['attempt']` is 1 based, but the `self.attempts` is 0 based? This will help with the person who were as confused as I am.